src/asmatrix.c

```diff
@@ -27,15 +48,30 @@ SEXP asmatrix_logical(SEXP dt, R_xlen_t matlen, R_xlen_t n, R_xlen_t rncolnum) {
 SEXP asmatrix_integer(SEXP dt, R_xlen_t matlen, R_xlen_t n, R_xlen_t rncolnum) {
   SEXP mat = PROTECT(allocVector(INTSXP, matlen)); // output matrix
   int *pmat, *pcol; // pointers to casted R objects
-  R_xlen_t vecIdx = 0; // counter to track place in vector underlying matrix
-  
-  // Iterate through dt and copy into mat 
   pmat = INTEGER(mat);
-  for (R_xlen_t jj = 0; jj < xlength(dt); jj++) {
-    if (jj == rncolnum) continue; // skip rownames. TODO: make sure this does slow down because of branch prediction
-    pcol = INTEGER(VECTOR_ELT(dt, jj));
-    memcpy(pmat + vecIdx, pcol, sizeof(int)*n);
-    vecIdx += n;
+
+  // Set up parallel OMP chunk
+  R_xlen_t dtncol = xlength(dt);
+  R_xlen_t chunksize = dtncol/getDTthreads() + 1; 
+  #pragma omp parallel num_threads(getDTthreads()) private(pcol)
+  { 
+    // Determine which columns in DT we're iterating through on this thread
+    int threadnum = omp_get_thread_num();
+    R_xlen_t startcol = threadnum * chunksize;
+    R_xlen_t endcol = startcol + chunksize - 1;
+    if (endcol >= dtncol) endcol = xlength(dt) - 1;
+    
+    // Determine where to fill in the matrix vector
+    R_xlen_t vecIdx = startcol * n;
+    if (rncolnum < startcol) vecIdx -= n;
+    
+    // Iterate through columns in DT, copying the contents to the matrix column
+    for (R_xlen_t jj = startcol; jj <= endcol; jj++) {
+      if (jj == rncolnum) continue; // skip rownames.
+      pcol = INTEGER(VECTOR_ELT(dt, jj));
```

Comment:
create array of pointers outside of parallel region and here use plain C objects